[Client] Organize component structure by Atomic Design principles Yob branch

Eyob-ts opened the pull request.

aimalamiri commented on Clients/src/components/atoms/Button.vue:
```diff
@@ -0,0 +1,119 @@
+          <circle class="opacity-25" cx="12" cy="12" r="10" stroke="currentColor" stroke-width="4"></circle>
+          <path class="opacity-75" fill="currentColor" d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"></path>
+        </svg>
+      </span>
```
Nice job, but don't you think this icon can go to its own component as an atom? Let's move it.

aimalamiri commented on Clients/src/components/atoms/Button.vue:
```diff
@@ -0,0 +1,119 @@
+      <!-- Button Content -->
+      <span 
+        class="flex items-center justify-center transition-opacity duration-200"
+        :class="{'opacity-0': loading, 'opacity-100': !loading}"
```
We can do something like:  

:class="`flex items-center justify-center transition-opacity duration-200 ${loading ? 'opacity-0' : 'opacity-100'}`"

aimalamiri commented on Clients/src/components/atoms/Button.vue:
```diff
@@ -0,0 +1,119 @@
+      <span
+        v-if="!disabled && !loading"
+        class="absolute inset-0 bg-white opacity-0 group-hover:opacity-10 group-active:opacity-20 transition-opacity duration-300"
+      ></span>
```
These are a lot of tags, and we can extract them to atoms, and reuse them in here, so the button will become a molecule at the end of the day.

aimalamiri commented on Clients/src/components/atoms/Button.vue:
```diff
@@ -0,0 +1,119 @@
+    secondary: 'bg-gray-100 text-gray-800 hover:bg-gray-200 dark:bg-gray-700 dark:text-white dark:hover:bg-gray-600 focus:ring-gray-300',
+    ghost: 'bg-transparent text-red-600 hover:bg-red-50 focus:ring-red-200',
+    danger: 'bg-red-100 text-red-700 hover:bg-red-200 focus:ring-red-300',
+    success: 'bg-green-100 text-green-700 hover:bg-green-200 focus:ring-green-300'
```
Nice job!

aimalamiri commented on Clients/src/components/atoms/PlacehoderImage.vue:
```diff
@@ -0,0 +1,13 @@
+        <svg class="w-16 h-16 text-gray-400 mx-auto mb-4 " fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
+            <path fill-rule="evenodd" d="M10 2a8 8 0 100 16 8 8 0 000-16zm-1.5 4a.5.5 0 01.5-.5h3a.5.5 0 010 1h-3a.5.5 0 01-.5-.5zM10 12a1.5 1.5 0 100-3 1.5 1.5 0 000 3z" clip-rule="evenodd"/>
+
+        </svg>
```
Again, this can go to a different component.

aimalamiri commented on Clients/src/components/atoms/PlaceholderImage.vue:
```diff
@@ -0,0 +1,10 @@
+      <div class="text-center">
+        <svg class="w-16 h-16 text-gray-400 mx-auto mb-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
+          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M7 4v16M17 4v16M3 8h4m10 0h4M3 12h18M3 16h4m10 0h4M4 20h16a1 1 0 001-1V5a1 1 0 00-1-1H4a1 1 0 00-1 1v14a1 1 0 001 1z" />
+        </svg>
```
Please extract all icons to its own component.

aimalamiri commented on Clients/src/components/molecules/MovieCard.vue:
```diff
@@ -0,0 +1,52 @@
+      <PlaceholderImage v-else/>
+
+        <!-- Movie Info Overlay -->
+         <div class="absolute inset-0 bg-gradient-to-t from-black/80 to-transparent opacity-0 group-hover:opacity-100 duration-300">
```
We can put this div into a new component as well.

aimalamiri commented on Clients/src/components/molecules/MovieList.vue:
```diff
@@ -0,0 +1,67 @@
+            />
+        </div>
+
+        <!-- No Results -->
```
Useless comment, let's remove it and make it cleaner

aimalamiri commented on Clients/src/components/ui/Button.vue:
```diff
@@ -0,0 +1,55 @@
+        ? 'bg-gradient-to-br from-red-700 to-black/20 text-white hover:from-red-600 hover:to-red-700 focus:ring-red-400 shadow-white/5'
+        : variant === 'secondary'
+        ? 'bg-gray-100 dark:bg-gray-700 text-gray-800 dark:text-white hover:bg-gray-200 dark:hover:bg-gray-600 focus:ring-gray-300'
+        : 'bg-transparent text-red-600 dark:text-red-400 hover:bg-red-50 dark:hover:bg-gray-700 focus:ring-red-200',
```
Let's put this into variables like you did in above

aimalamiri commented on Clients/src/components/ui/Input.vue:
```diff
@@ -0,0 +1,50 @@
+        focus:ring-2 
+        transition duration-700
+      "
+    />
```
These can be two atoms, and one molecule

aimalamiri commented on Clients/src/pages/auth/Login.vue:
```diff
@@ -0,0 +1,144 @@
+</script>
+
+<style scoped>
+.centered-container {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  height: 100vh;
+}
+
+.auth-form {
+  max-width: 400px;
+  padding: 20px;
+  background-color: #f9f9f9;
+  border: 1px solid #ddd;
+  border-radius: 10px;
+  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
+}
+
+.input-field {
+  width: 100%;
+  padding: 10px;
+  margin-bottom: 20px;
+  border: 1px solid #ccc;
+  border-radius: 5px;
+}
+
+.btn-primary {
+  background-color: #4CAF50;
+  color: #fff;
+  padding: 10px 20px;
+  border: none;
+  border-radius: 5px;
+  cursor: pointer;
+}
```
We have tailwindcss, do we still need this much custom css?

aimalamiri commented on Clients/src/pages/auth/Register.vue:
```diff
@@ -0,0 +1,106 @@
+      v-model="form.password"
+        label="Password"
+        type="password"
+        placeholder="••••••••"
```
Great idea, but don't you think it will be confusing a bit for the user, and he'll think there is already a password entered?
Instead we can write "Enter your password" in the placeholder.

aimalamiri commented on Clients/src/pages/auth/Login.vue:
```diff
@@ -0,0 +1,144 @@
+      toast.error(authStore.error?.message || 'Login failed')
+    }
+  } catch (error) {
+    console.error('Login error:', error)
```
Good for the development, but let's remove it from here.

aimalamiri commented on Clients/src/pages/movies/SearchPage.vue:
```diff
@@ -0,0 +1,77 @@
+            variant="primary"
+          />
+        </div>
+        <!-- Search Components -->
+        <div class="flex gap-4 items-end">
+          <Input
+            v-model="searchQuery"
+            @keyup.enter="handleSearch"
+            label="Search for a movie"
+            placeholder="Enter movie title"
+            class="flex-1"
+          />
+          <Button
+            @click="handleSearch"
+            :disabled="!searchQuery.trim() || movieStore.isLoading"
+            :loading="movieStore.isLoading"
+            label="Search"
+            variant="primary"
+          />
+        </div>
```
Great job, but why don't we extract this into a SearchComponent instead of commenting it in here?

aimalamiri commented on Clients/src/pages/movies/SearchPage.vue:
```diff
@@ -0,0 +1,77 @@
+      </div>
+    </header>
+
+    <!-- Main Content -->
+    <div class="max-w-7xl mx-auto py-6 sm:px-6 lg:px-8">
+      <div class="px-4 py-6 sm:px-0 bg-gradient-to-br from-black to-red-500">
+        <MovieList :search-query="searchQuery" />
+      </div>
+    </div>
```
Same here

aimalamiri commented on Clients/src/services/auth.api.js:
```diff
@@ -0,0 +1,41 @@
+import axios from 'axios';
+
+const API_URL = "http://127.0.0.1:8000/api";
```
This can be a .env variable, so in the deployment server we won't need to edit this code file. And also in another dev computer as well.

aimalamiri commented on Clients/src/services/auth.api.js:
```diff
@@ -0,0 +1,41 @@
+    },
+    (error) => {
+        if (error.response && error.response.status === 401) {
+            // Handle unauthorized access
+            localStorage.removeItem('token');
+            window.location.href = '/login'; // Redirect to login page
```
It's okay to use Ai for generating the code, but the please make sure to check the code and clean it up. So let's remove these comments.

aimalamiri commented on Clients/src/services/movie.api.js:
```diff
@@ -0,0 +1,58 @@
+import axios from 'axios';
+
+const apiClient = axios.create({
+  baseURL: 'https://imdb8.p.rapidapi.com',
+  headers: {
+    'x-rapidapi-key': '68261eb261msh44ee9981e3b673cp1e6ab3jsn0addcc419975',
+    'x-rapidapi-host': 'imdb8.p.rapidapi.com'
+  }
```
Seems like we need 3 more environment variables, right?

aimalamiri commented on Clients/src/store/auth.store.js:
```diff
@@ -0,0 +1,64 @@
+        async register(credentials) {
+            try {
+                const response = await axiosInstance.post('/register', credentials);
+                this.token = response.data.token;
+                this.user = response.data.user;
+                localStorage.setItem('token', this.token);
+                axiosInstance.defaults.headers.common['Authorization'] = `Bearer ${this.token}`;
+                return { success: true, user: this.user };
+            } catch (error) {
+                this.error = error.response?.data || { message: 'Registration failed' };
+                console.error('Registration failed:', error);
+                return { success: false, error: this.error };
+            }
+        },
+        async login(credentials) {
+            try {
+                const response = await axiosInstance.post('/login', credentials);
+                this.token = response.data.token;
+                this.user = response.data.user;
+                localStorage.setItem('token', this.token);
+                axiosInstance.defaults.headers.common['Authorization'] = `Bearer ${this.token}`;
+                return { success: true, user: this.user };
+            } catch (error) {
+                this.error = error.response?.data || { message: 'Login failed' };
+                return { success: false, error: this.error };
+            }
+        },
+        async logout() {
+            try {
+                await axiosInstance.post('/logout');
+                this.token = null;
+                this.user = null;
+                localStorage.removeItem('token');
+                delete axiosInstance.defaults.headers.common['Authorization'];
+                return {success: true};
+            } catch (error) {
+                this.error = error.response?.data || { message: 'Logout failed' };
+                console.error('Logout failed:', error);
+                return {success: false, error: this.error};
+            }
```
Do you see the common pattern between these functions? all of them use almost the same repeated code except a few changes like endpoint, let's clear it up by extracting it to a new function and reuse it for all three.

aimalamiri commented on Server/config/cors.php:
```diff
@@ -1,34 +1,12 @@
-
-    'allowed_origins' => ['*'],
-
+    'allowed_origins' => ['http://localhost:5173', 'http://localhost:3000'],
```
Please put this int an env variable, otherwise it'll be required to change the code file for deploying the app.

aimalamiri commented on Server/config/cors.php:
```diff
@@ -1,34 +1,12 @@
-
-    'supports_credentials' => false,
-
+    'supports_credentials' => true,
```
Can you explain this? what it is and what it does?

Eyob-ts replied:
@aimalamiri 
This is a CORS configuration in laravel. It controls which external domain (like our Vue.js frontend) are allowed to access the backend API

What it Does :- Allow the browser to send cookies and authentication header (like session tokens) when our Vue app calls the Laravel API from a different domain

aimalamiri reviewed (changes requested):
Hi @Eyob-ts, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

![AlmostThereReachGIF](https://user-images.githubusercontent.com/34779928/191347504-9f70659c-a01f-48a1-ad46-b8722d46ffca.gif)

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

Eyob-ts commented:
@aimalamiri  Wow I can't thank you enough for this incredibly thorough review! As my first Vue.js project, I know my code had ... rough edges (to Put it kindly : ).  your feedback not only fixes issues but teaches me better patterns. Really appreciate you taking time to help me level up.  This is exactly the mentorship I needed!

I'll make all the requested changes and implement your suggestions.

Eyob-ts commented:
@aimalamiri  I've addressed all feedback from your review in the latest push
Updated Input component with improved styling and accessibility features.
- Replaced deprecated PlaceholderImage component with a new implementation.
- Removed unused MovieCard and MovieList components to streamline the codebase.
- Enhanced NotFound page with a new button component for navigation.
- Updated Login and Register pages to use FormField component for better structure.
- Improved SearchPage layout and integrated a new SearchBar component for movie searches.
- Refined authentication store to handle token management and error messages more effectively.
- Adjusted API service to utilize environment variables for configuration.
